src/internal/FluencePeer.ts

```diff
@@ -349,79 +358,94 @@ export class FluencePeer {
+                }),
                 filterExpiredParticles(this._expireParticle.bind(this)),
             )
-            .subscribe((p) => {
+            .subscribe((item) => {
+                const p = item.particle;
                 let particlesQueue = this._particleQueues.get(p.id);
 
                 if (!particlesQueue) {
                     particlesQueue = this._createParticlesProcessingQueue();
                     this._particleQueues.set(p.id, particlesQueue);
 
                     const timeout = setTimeout(() => {
-                        this._expireParticle(p.id);
+                        this._expireParticle(item);
                     }, p.actualTtl());
 
                     this._timeouts.push(timeout);
                 }
 
-                particlesQueue.next(p);
+                particlesQueue.next(item);
             });
 
-        this._outgoingParticles.subscribe((p) => {
-            this._connection.sendParticle(p);
+        this._outgoingParticles.subscribe(async (item) => {
+            await this._connection.sendParticle(item.particle);
+            item.onStageChange({ stage: 'sent' });
         });
     }
 
-    private _expireParticle(particleId: string) {
+    private _expireParticle(item: ParticleQueueItem) {
+        const particleId = item.particle.id;
         log.debug(`particle ${particleId} has expired. Deleting particle-related queues and handlers`);
 
         this._particleQueues.delete(particleId);
-        const timeoutHandler = this._timeoutHandlers.get(particleId);
-        if (timeoutHandler) {
-            timeoutHandler();
-        }
         this._particleSpecificHandlers.delete(particleId);
-        this._timeoutHandlers.delete(particleId);
+
+        item.onStageChange({ stage: 'expired' });
     }
 
     private _createParticlesProcessingQueue() {
-        let particlesQueue = new Subject<Particle>();
+        let particlesQueue = new Subject<ParticleQueueItem>();
         let prevData: Uint8Array = Buffer.from([]);
 
         particlesQueue
             .pipe(
                 // force new line
                 filterExpiredParticles(this._expireParticle.bind(this)),
             )
-            .subscribe((x) => {
-                const result = runInterpreter(this.getStatus().peerId, this._interpreter, x, prevData);
+            .subscribe((item) => {
+                const particle = item.particle;
+                const result = runInterpreter(this.getStatus().peerId, this._interpreter, particle, prevData);
+
+                // Do not continue if there was an error in particle interpretation
+                if (isInterpretationSuccessful(result)) {
+                    item.onStageChange({ stage: 'interpreterError', errorMessage: result.errorMessage });
+                    return;
+                }
+
+                setTimeout(() => {
+                    item.onStageChange({ stage: 'interpreted' });
+                }, 0);
 
                 prevData = Buffer.from(result.data);
```

Comment:
It's not `prevData`, it's `newData`. 

The result of this naming is confusing:
```
newParticle.data = prevData;
```

Reply:
Well, it's new data and prev data at the same time. Right after interpretation the data becomes new. But when it get's interpreted next time it's already prevData

Reply:
Just rename it to `newData`, please.